wordpress-mobile/WordPress-Android
# Blaze: Implement promote with blaze web view flow

Parent #18001 

This PR launches the Blaze flow inside of a webview.

Post List Light Header | Post List Dark Header
-- | --
<img width="250" height="500" alt="Alt desc" src="https://user-images.githubusercontent.com/506707/222023512-2b25a4e8-342e-4c4e-8532-d69e07e87b2a.png"> | <img width="250" height="500" alt="Alt desc" src="https://user-images.githubusercontent.com/506707/222023527-e99c6a54-3681-49e1-b9a0-be949d0f2f3d.png">


Created the following to manage the Blaze create campaign webview flow. 
Includes: 
- Created `BlazeWebViewFragment`
- Created `BlazeWebViewViewModel`
- Created `BlazeWebClient`
- Created `BlazeActionEvent` for communication from the BlazeWebViewViewModel back to the BlazeWebViewFragment. For situation for finishing activity on error or flow ended
- Created and added the `BlazeWebViewHeaderUiState` and `BlazeWebViewContentUiState` states to the BlazeUiStates class. I toyed with using "model" vs. "UiState" suffix and am willing to rename as many times as necessary
- Do not allow the user to "back" out of the flow, they must use the cancel button
- "Blaze" is aligned to the left in the top header
- Tracking for blaze flow: started, canceled, error, and completed
 
Refactored: 
BlazeUiState:
 - Got rid of non-native uiStates to reflect the fact that we do not manage the "webView" flow like we do for a native flow.
 - Added "WebViewScreen" as a state
 - Added "page" as state (this was added as an entry point in the tech spec last week)
 - Added a getSource() to BlazeViewModel so I can grab it and pass along to BlazeWebViewViewModel )
 - Any errors encountered will push finishing the activity; push the user back to their entry point

BlazeParentActivity
- Launch the web view flow when the state changes to WebViewScreen 


BlazeViewModel
- Added the update for uiState onPromoteWithBlazeClicked (yes, I know this is a button click will be updated, but I wanted to make sure the showNextScreen worked)
- Adjusted the `showNextScreen` fun
- Added getSource() so I can grab the value in the BlazeWebView flow 

**Outstanding items - These will be handled in separate PRs**
- Page entry points handling
- Unit tests
- Compose look/feel of the header with regards to text placement/color (disabled doesn't look disabled yet) and webview in dark mode. This will go in a separate PR so Shaun can review as well.

**Known Issues**
- Disabling of the Cancel header action while payment is processing. This is being addressed on the server side.
- Tapping a "reader" external link launches the app reader, not in an external browser. Not sure how to approach this one yet.

**To test:**
**Setup**
- Install and launch the app
- Navigate to Me -> App Settings -> Privacy Setting
- Enable collection information
- Navigate to Me -> App Settings -> Debug Setting
- Enable Blaze
- Restart the app

**Test Dashboard Entry Point**
- Navigate to My SIte tab
- Tap on the Promote With Blaze card
- Tap on the Blaze a Post Now button within the overlay
- ✅ Verify you are launched into the Blaze flow post lists
- ✅ Verify logs contain:  🔵 Tracked: blaze_flow_started, Properties: {"source":"dashboard_card"}
- Tap the cancel button in the header
- ✅ Verify you are taken out of the flow 
- ✅ Verify logs contain: 🔵 Tracked: blaze_flow_canceled, Properties: {"source":"dashboard_card","current_step":"posts_list"}
- Enter the flow again: Tap on the Promote With Blaze card > Tap on the Blaze a Post Now button within the overlay
- Select a post from the list and tap Promote
- Tap the cancel button in the header Appearance view
- ✅ Verify you are taken out of the flow 
- ✅ Verify logs contain: 🔵 Tracked: blaze_flow_canceled, Properties: {"source":"dashboard_card","current_step":"step_1"}

- Enter the flow again: Tap on the Promote With Blaze card > Tap on the Blaze a Post Now button within the overlay
- Select a post from the list and tap Promote
- Fill out the Appearance information and tap Next
- Tap the cancel button in the header Audience view
- ✅ Verify you are taken out of the flow 
- ✅ Verify logs contain: 🔵 Tracked: blaze_flow_canceled, Properties: {"source":"dashboard_card","current_step":"step_2"}

- Enter the flow again: Tap on the Promote With Blaze card > Tap on the Blaze a Post Now button within the overlay
- Select a post from the list and tap Promote
- Fill out the Appearance information and tap Next
- Fill out the Audience information and tap Next
- Tap the cancel button in the header on the Payment view
- ✅ Verify you are taken out of the flow 
- ✅ Verify logs contain: 🔵 Tracked: blaze_flow_canceled, Properties: {"source":"dashboard_card","current_step":"step_3"}

- Enter the flow again: Tap on the Promote With Blaze card > Tap on the Blaze a Post Now button within the overlay
- Select a post from the list and tap Promote
- Fill out the Appearance information and tap Next (Change the title and snippet to DO NOT APPROVE - TESTING)
- Fill out the Audience information and tap Next
- Fill out the Payment information and tap Save & Submit
- Note that the cancel button will still be visible and clickable, please do not click it. Need server change.
- Wait for the process to complete
-  ✅ Verify the header action button changed to "done"
- DO NOT TAP DONE YET (unless you want to go back through the process again ;) )
- Tap the TermsConditions link
-  ✅ Verify an external browser is launched
- Navigate back to the app
- Tap the View Campaigns button
- Tap the Header "Done" button
- ✅ Verify logs contain: 🔵 Tracked: blaze_flow_completed, Properties: {"source":"dashboard_card"}

**Test Post Entry**
- Navigate to My SIte tab
- Tap on the Posts menu item or ribbon action
- Tap on More > Promote with Blaze
- ✅ Verify you are launched into the Blaze flow at the Appearance page for the post you selected
- ✅ Verify logs contain:  🔵 Tracked: blaze_flow_started, Properties: {"source":"posts_list"}
- Tap the cancel button in the header
- ✅ Verify you are taken out of the flow 
- ✅ Verify logs contain: 🔵 Tracked: blaze_flow_canceled, Properties: {"source":"posts_list","current_step":"step_1"}
* this is a streamlined test, as the only thing that changes from the first set of tests is the source. Feel free to go through the complete flow if you want to verify this.

## Regression Notes
1. Potential unintended areas of impact
Blaze flow does not work properly

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing

3. What automated tests I added (or what prevented me from doing so)
Separate PR

PR submission checklist:
